Add GeniePlus Harvester

demiankatz opened the pull request:
TODO
- [x] Merge in work from #98 when completed

demiankatz committed:
Work in progress on GeniePlus harvest.

demiankatz committed:
More progress.

demiankatz committed:
Improve alias.

demiankatz committed:
Progress on implementation.

demiankatz committed:
Progress on MARC formatting.

demiankatz committed:
Add ID injection.

demiankatz committed:
Add sample configs.

demiankatz committed:
Add date range support.

demiankatz committed:
XML record sanitization.

demiankatz committed:
Add 852 support; fix date conversion.

demiankatz committed:
Fix styles.

demiankatz committed:
Make PHPStan happy.

demiankatz committed:
Merge remote-tracking branch 'origin/dev' into genieplus

demiankatz committed:
Merge remote-tracking branch 'origin/dev' into genieplus

demiankatz commented:
@EreMaijala, please consider this a draft for now, but I wanted to share my work in progress in case you have any feedback.

demiankatz committed:
Fix typo.

EreMaijala commented on conf/datasources.ini.sample:
```diff
@@ -187,3 +187,33 @@ institution = SierraLib
+; Sample configuration for GeniePlus REST API
+[GeniePlus]
+type = GeniePlus
+genieplusDatabase = "[api_database_name]"
```
This one as well as `genieplusTemplate` have a lowercase p, should be unified.

demiankatz replied:
Fixed!

EreMaijala commented on src/RecordManager/Base/Harvest/GeniePlus.php:
```diff
@@ -0,0 +1,682 @@
+        $this->password = $settings['geniePlusPassword'];
+        $this->batchSize = $settings['batchSize'] ?? 100;
+        $this->idField = $settings['geniePlusIdField'] ?? 'UniqRecNum';
+        $this->marcField = $settings['MarcRecord'] ?? 'MarcRecord';
```
Should this be `$settings['geniePlusMarcField']`?

demiankatz replied:
Fixed!

EreMaijala commented on src/RecordManager/Base/Harvest/GeniePlus.php:
```diff
@@ -0,0 +1,682 @@
+            $replacementCount
+        );
+        if ($replacementCount > 0) {
+            $this->infoMsg("Replaced $replacementCount bad chars in record $id");
```
infoMsg may be fine, but consider warningMsg/debugMsg as well depending on how frequent occurrence this is and how much bearing it has on the resulting records.

demiankatz replied:
Yes, I think warning is better -- changed (along with merging in #98).

EreMaijala commented:
@demiankatz Looks pretty good to me already. I glanced through the code and added a couple of observations.

demiankatz committed:
Merge branch 'dev' into genieplus

demiankatz committed:
PR feedback fixes; updated comments.

demiankatz commented:
Thanks, @EreMaijala, I think we're in pretty good shape now -- I've addressed all of your feedback. Feel free to merge this if you think it's ready, or to leave it open if you want to give it a few days to be sure I don't make more changes. I think I've met all of my immediate needs, but it's possible I might come up with a tweak or two as I test further.

EreMaijala reviewed (approved):
I can't actually test this, but looks good to me!
